clinica/utils/input_files.py

```diff
@@ -194,12 +199,54 @@
 # T1-Volume
 
 
+def aggregator(func):
+    """If the decorated function receives iterable arguments,
+    this decorator will call the decorated function for each
+    value in the iterable and aggregate the results in a list.
+    This works only if the iterables provided have the same length.
+    Arguments lefts as non-iterable will be repeated.
+    """
+
+    @functools.wraps(func)
+    def wrapper_aggregator(*args, **kwargs):
+        arg_sizes = [len(arg) for arg in args if isinstance(arg, Iterable)]
+        arg_sizes += [
+            len(arg) for k, arg in kwargs.items() if isinstance(arg, Iterable)
+        ]
+        if len(np.unique(arg_sizes)) > 1:
+            raise ValueError(f"Arguments must have the same length.")
+        if len(arg_sizes) == 0:
+            return func(*args, **kwargs)
+        arg_size = arg_sizes[0]
+        new_args = []
+        for arg in args:
+            if not isinstance(arg, Iterable):
+                new_args.append((arg,) * arg_size)
+            else:
+                new_args.append(arg)
+        new_kwargs = [{} for _ in range(arg_size)]
+        for k, arg in kwargs.items():
+            for i in range(len(new_kwargs)):
+                if not isinstance(arg, Iterable):
+                    new_kwargs[i][k] = arg
+                else:
+                    new_kwargs[i][k] = arg[i]
+        if len(new_args) == 0:
+            return [func(**x) for x in new_kwargs]
+        elif len(new_kwargs) == 0:
+            return [func(*x) for x in zip(*new_args)]
+        return [func(*x, **y) for x, y in zip(zip(*new_args), new_kwargs)]
```

Comment:
This is painful to read without comments to dissociate the cases covered by this decorator and their intermediate steps.

Reply:
Yes, I'm not proud of this function and I totally get that it is very painful to read.

I initially thought there was a more elegant/clever way to do this but couldn't find one. If you have suggestions/ideas I'll take them.

To try answering your points, I added a few things in [80b8d88](https://github.com/aramis-lab/clinica/pull/760/commits/80b8d88dfdb63936783ac0004838aaf58ebd9e00):

- Basic unit tests for this function
- Examples and basic explanation of why we need this
- A few comments to break and explain the chunks of code

Hopefully this is enough to understand what this decorator does and how to use it.